libs/core/src/logging.cpp

```diff
@@ -0,0 +1,199 @@
+//
+//   Unless required by applicable law or agreed to in writing, software
+//   distributed under the License is distributed on an "AS IS" BASIS,
+//   WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+//   See the License for the specific language governing permissions and
+//   limitations under the License.
+//
+//------------------------------------------------------------------------------
+
+#include "core/logging.hpp"
+#include "core/mutex.hpp"
+
+#include "spdlog/sinks/stdout_color_sinks.h"
+#include "spdlog/spdlog.h"
+
+#include <unordered_map>
+
+namespace fetch {
+namespace {
+
+class LogRegistry
+{
+public:
+  // Construction / Destruction
+  LogRegistry();
+  LogRegistry(LogRegistry const &) = delete;
+  LogRegistry(LogRegistry &&)      = delete;
+  ~LogRegistry()                   = default;
+
+  void        Log(LogLevel level, char const *name, std::string &&message);
+  void        SetLevel(char const *name, LogLevel level);
+  LogLevelMap GetLogLevelMap();
+
+  // Operators
+  LogRegistry &operator=(LogRegistry const &) = delete;
+  LogRegistry &operator=(LogRegistry &&) = delete;
+
+private:
+  using Logger    = spdlog::logger;
+  using LoggerPtr = std::shared_ptr<Logger>;
+  using Registry  = std::unordered_map<std::string, LoggerPtr>;
+  using Mutex     = std::mutex;
+
+  Logger &GetLogger(char const *name);
+
+  Mutex    lock_;
+  Registry registry_;
+};
+
+constexpr LogLevel DEFAULT_LEVEL = LogLevel::INFO;
+
+LogRegistry registry_;
+
+LogLevel ConvertToLevel(spdlog::level::level_enum level)
+{
+  LogLevel new_level = LogLevel::INFO;
+
+  switch (level)
+  {
+  case spdlog::level::trace:
+    new_level = LogLevel::TRACE;
+    break;
+  case spdlog::level::debug:
+    new_level = LogLevel::DEBUG;
+    break;
+  case spdlog::level::info:
+    new_level = LogLevel::INFO;
+    break;
+  case spdlog::level::warn:
+    new_level = LogLevel::WARNING;
+    break;
+  case spdlog::level::err:
+    new_level = LogLevel::ERROR;
+    break;
+  case spdlog::level::critical:
+    new_level = LogLevel::CRITICAL;
+    break;
+  case spdlog::level::off:
+    break;
+  }
+
+  return new_level;
+}
+
+spdlog::level::level_enum ConvertFromLevel(LogLevel level)
+{
+  spdlog::level::level_enum new_level = spdlog::level::info;
+
+  switch (level)
+  {
+  case LogLevel::TRACE:
+    new_level = spdlog::level::trace;
+    break;
+  case LogLevel::DEBUG:
+    new_level = spdlog::level::debug;
+    break;
+  case LogLevel::INFO:
+    new_level = spdlog::level::info;
+    break;
+  case LogLevel::WARNING:
+    new_level = spdlog::level::warn;
+    break;
+  case LogLevel::ERROR:
+    new_level = spdlog::level::err;
+    break;
+  case LogLevel::CRITICAL:
+    new_level = spdlog::level::critical;
+    break;
+  }
+
+  return new_level;
+}
+
+LogRegistry::LogRegistry()
+{
+  spdlog::set_level(
+      spdlog::level::trace);  // this should be kept in sync with the compilation level
```

Comment:
presumably debug mode compiles all levels and release some subset?

Reply:
Current the same behaviour is kept. i.e. info messages and "above" are compiled into the code for all modes.

Stage 2. Will be to alter this so that the default debug mode is configuration but will default to debug messages I think. There is a trade off here between slow down in debug mode and usefulness to be able to switch any mode on.

This particular line is the global runtime log filter level. When I change the compile time configuration I will make sure this is matched here. This sort of gives the user a runtime API to query what the compiled log level is. (that was the thinking anyway)